src/util/file_watch.c

```diff
@@ -0,0 +1,331 @@
+    if (use_inotify) {
+        ret = try_inotify(fw_ctx);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_MINOR_FAILURE, "Falling back to polling\n");
+            use_inotify = false;
+        }
+    }
+
+    if (!use_inotify) {
+        ret = fw_watch_file_fallback(mem_ctx, fw_ctx);
+    }
+
+    return ret;
+}
+
+
+static void missing_file(struct tevent_context *ev,
+                                struct tevent_timer *te,
+                                struct timeval tv, void *data)
+{
+    int ret;
+    struct file_watch_ctx *fw_ctx = talloc_get_type(data, struct file_watch_ctx);
+
+    ret = watch_file(fw_ctx, fw_ctx);
+    if (ret == EOK) {
+        fw_ctx->fn(fw_ctx->filename, fw_ctx->fn_arg);
+    } else if (ret == ENOENT) {
+        DEBUG(SSSDBG_MINOR_FAILURE,
+              "%s missing. Waiting for it to appear.\n",
+              fw_ctx->filename);
+        tv = tevent_timeval_current_ofs(MISSING_FILE_POLL_TIME, 0);
+        te = tevent_add_timer(fw_ctx->ev, fw_ctx, tv, missing_file, fw_ctx);
+        if (te == NULL) {
+            DEBUG(SSSDBG_FATAL_FAILURE,
+                  "tevent_add_timer failed. %s will be ignored.\n",
+                  fw_ctx->filename);
+        }
+    } else {
+        DEBUG(SSSDBG_FATAL_FAILURE,
+              "watch_file failed. %s will be ignored.\n",
+              fw_ctx->filename);
+    }
+}
+
+
+struct file_watch_ctx *fw_watch_file(TALLOC_CTX *mem_ctx,
+                                     struct tevent_context *ev,
+                                     const char *filename,
+                                     bool use_inotify,
+                                     fw_fn fn,
+                                     void *fn_arg)
+{
+    int ret;
+    struct timeval tv;
+    struct tevent_timer *te;
+    struct file_watch_ctx *fw_ctx;
+
+    if (ev == NULL || filename == NULL || fn == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "Invalid parameter\n");
+        return NULL;
+    }
+
+    fw_ctx = talloc_zero(mem_ctx, struct file_watch_ctx);
+    if (fw_ctx == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "Failed to allocate the context\n");
+        return NULL;
+    }
+
+    fw_ctx->ev = ev;
+    fw_ctx->use_inotify = use_inotify;
+    fw_ctx->fn = fn;
+    fw_ctx->fn_arg = fn_arg;
+    fw_ctx->filename = talloc_strdup(fw_ctx, filename);
+    if (fw_ctx->filename == NULL) {
+        DEBUG(SSSDBG_CRIT_FAILURE, "talloc_strdup() failed\n");
+        ret = ENOMEM;
+        goto done;
+    }
+
+    /* Watch for changes to the requested file */
+    ret = watch_file(fw_ctx, fw_ctx);
+    if (ret == ENOENT) {
+        tv = tevent_timeval_current_ofs(MISSING_FILE_POLL_TIME, 0);
+        te = tevent_add_timer(fw_ctx->ev, fw_ctx, tv, missing_file, fw_ctx);
+        if (te == NULL) {
+            DEBUG(SSSDBG_FATAL_FAILURE, "%s will be ignored\n", filename);
+        }
+    }
+
+done:
+    if (ret != EOK && ret != ENOENT) {
+        talloc_free(fw_ctx);
+        fw_ctx = NULL;
+    }
+
```

Comment:
Is there anything to cleanup in `file_watch_ctx` destructor?

Reply:
It is enough to `talloc_free()` it.

Reply:
Will this destroy inotify watch (didn't check the code).

Reply:
In `inotify.c:484`, a destructor is set to destroy the inotify watch. This is the only part I'm aware of that needs a specific clean-up.

`talloc_free()` will free all the `file_watch_ctx` memory and trigger `inotify.c`'s  destructor. However, some extra work might be required if timers (when polling) are pending.

Reply:
It seems that timers are created using the parent menroy context, so they are treated correctly.

Reply:
See https://github.com/SSSD/sssd/pull/6421#discussion_r1054848662

Reply:
Now `parent_ctx` was removed and timers are linked to the `file_watch_ctx`.